File: go/0212-word-search-ii.go
type TrieNode struct {
    children map[byte]*TrieNode
    isWord bool
    refs int
}

func (this *TrieNode) addWord(word string) {
    cur := this
    cur.refs += 1
    for c := 0; c < len(word); c++ {
        if _, ok := cur.children[word[c]]; !ok {
            cur.children[word[c]] = &TrieNode{children: make(map[byte]*TrieNode)}
        }
        cur = cur.children[word[c]]
        cur.refs += 1
    }
    cur.isWord = true
}

func (this *TrieNode) removeWord(word string) {
    cur := this
    cur.refs += 1
    for c := 0; c < len(word); c++ {
        if _, ok := cur.children[word[c]]; ok {
            cur = cur.children[word[c]];
            cur.refs -= 1
        }
    }
}

func findWords(board [][]byte, words []string) []string {
    root := &TrieNode{children: make(map[byte]*TrieNode)}
    for _, w := range words {
        root.addWord(w);
    }
    
    ROWS, COLS := len(board), len(board[0])
    res := make(map[string]bool)
    visit := make(map[int]bool)
    
    var dfs func(int, int, *TrieNode, string)
    dfs = func(r, c int, node *TrieNode, word string) {
        if
            r < 0 || r >= ROWS ||
            c < 0 || c >= COLS ||
            node.children[board[r][c]] == nil ||
            node.children[board[r][c]].refs < 1 ||
            visit[r*COLS + c] {
            return
        }
        
        visit[r*COLS + c] = true
        node = node.children[board[r][c]]
        word += string(board[r][c])
        if node.isWord {
            node.isWord = false
            res[word] = true
            root.removeWord(word)
        }
        
        dfs(r + 1, c, node, word)
        dfs(r - 1, c, node, word)
        dfs(r, c + 1, node, word)
        dfs(r, c - 1, node, word)
        visit[r*COLS + c] = false
    }
    
    for r := 0; r < ROWS; r++ {
        for c := 0; c < COLS; c++ {
            dfs(r, c, root, "")
        }
    }
    return list(res)
}

func list(mapping map[string]bool) []string {
    res := make([]string, 0, len(mapping))
    for key, _ := range mapping {
        res = append(res, key)
    }
    return res
}

File: typescript/0212-word-search-ii.ts
class TrieNode {
    children = new Array<string>(26);
    word = false;
}

class Trie {
    root = new TrieNode();

    add(word: string) {
        let current = this.root;
        for (let c of word) {
            if (!current.children[c]) current.children[c] = new TrieNode();
            current = current.children[c];
        }

        current.word = true;
    }
}

function findWords(board: string[][], words: string[]): string[] {
    const trie = new Trie();
    for (let word of words) {
        trie.add(word);
    }

    const rows = board.length;
    const cols = board[0].length;
    const resultSet = new Set<string>();

    function dfs(r: number, c: number, node: TrieNode, word: string) {
        if (r < 0 || r >= rows) return;
        if (c < 0 || c >= cols) return;
        if (board[r][c] === '.') return;

        const currentChar = board[r][c];
        const currentNode = node.children[currentChar];
        if (!currentNode) return;
        
        const currentWord = word + currentChar;
        if (currentNode.word) {
            resultSet.add(currentWord);
        }

        const directions = [
            [r - 1, c],
            [r, c + 1],
            [r + 1, c],
            [r, c - 1],
        ];
        for (let [nr, nc] of directions) {
            dfs(nr, nc, currentNode, currentWord);
        }

        board[r][c] = '.';
        board[r][c] = currentChar;
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            dfs(r, c, trie.root, '');
        }
    }

    return [...resultSet];
};
